### PR TITLE
rsx: Implement xform-constant-based instancing

### PR DESCRIPTION
**Pull request by kd-11**
We had other forms of instancing previously, but that was only done for index-base and vertex-base modifying setups. Some insomniac titles are using full xform-based instancing (load-matrix -> draw -> load-matrix -> draw) form of instancing which is not supported and is very performance intensive. This PR addresses that.

Fixes https://github.com/RPCS3/rpcs3/issues/10754

### TODO
- [x] Fix vulkan crashes on AMD due to vkCmdUpdateBuffer abuse.
- [x] Optimizations.
    - [x] Batch the updates on the barrier insertion side.
    - [x] Directly submit ranges to the backend to update. This avoids the indexed register penalty.
- [x] Implement the feature on OpenGL.
- [x] Investigate unusual coloration on some mesh instances (broken attributes??) [NOPE, lighting is borked on real PS3 too)
- [x] Investigate and fix performance drop on Vulkan

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/VKGSRender.cpp [hunk 1]
```diff
@@ -2290,6 +2292,23 @@ void VKGSRender::load_program_env()
...
+	if (transform_constants_size)
+	{
+		check_heap_status(VK_HEAP_CHECK_TRANSFORM_CONSTANTS_STORAGE);
+
+		buffer.reserve(transform_constants_size);
+		auto buf = buffer.data();
+
+		const auto constant_ids = (transform_constants_size == 8192)
+			? std::span<const u16>{}
+		: std::span<const u16>(m_vertex_prog->constant_ids);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKGSRender.cpp:2307]:
```suggestion
			: std::span<const u16>(m_vertex_prog->constant_ids);
```

**Comment by Linear524** [comment]
Can this fix #8442 ?
It is a long standing bug with grass cascades disappearing randomly...

**Comment by kd-11** [comment]
For some reason OpenGL is now almost twice as fast in scenes with heavy foliage. I need to optimize a little more by just going for a rebind when we can't do a hot-patch.

**Comment by kd-11** [comment]
There's a lot more to optimize after looking at profile traces.

File: rpcs3/Emu/RSX/Program/program_util.h [hunk 2]
```diff
@@ -57,4 +57,16 @@ namespace rsx
 		void set_dimension(texture_dimension_extended type, u32 index);
 		bool operator == (const vertex_program_texture_state& other) const;
 	};
+
+	struct VertexProgramBase
+	{
+		u32 id;
+		std::vector<u16> constant_ids;
+		bool has_indexed_constants;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Program/program_util.h:65]: uninitialized

File: rpcs3/Emu/RSX/Program/program_util.h [hunk 3]
```diff
@@ -57,4 +57,16 @@ namespace rsx
 		void set_dimension(texture_dimension_extended type, u32 index);
 		bool operator == (const vertex_program_texture_state& other) const;
 	};
+
+	struct VertexProgramBase
+	{
+		u32 id;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Program/program_util.h:63]: uninitialized

**Comment by kd-11** [comment]
Performance is fixed on AMD cards, but NVIDIA is struggling with a 60% performance loss. We can work around that by restructuring how vertex programs work which I'll have to do in parallel before this is merged.

**Comment by kd-11** [comment]
I'm going to merge this one as-is. The NVIDIA-specific optimization work will come later and will never be as fast as the approach taken here. I will try updating drivers and checking why the barriers are stalling so badly with Nsight and maybe reporting the issue to their driver team. Not much hope for quick action there though.

**Comment by kd-11** [comment]
I wonder if the spec violation is what is causing the performance to tank (buffer update in the middle of a renderpass). Maybe VK_KHR_dynamic_rendering can help out in that case. But every other "smart" trick I can think of including uploading a whole indirection table would require some kind of transfer op to patch the on-device lookup table which leaves us back at square 1.

**Comment by kd-11** [comment]
I need to think a bit more about this one. The only "fix" for NVIDIA here is to upgrade to VK1.3.

**Comment by JimScript** [comment]
I didn't think the fix for my issue would require work to be this..._extensive_, but thanks a ton kd-11!

**Comment by segevfiner** [comment]
@kd-11 Awesome! But is there a still a performance issue for NVIDIA or did you manage to find a workaround/fix? If not, I guess someone should open an issue to track that?

**Comment by homembarata** [comment]
Can confirm that the instanced foliage in Resistance is now showing up as it should. Great work, thanks a lot kd-11.

**Comment by kd-11** [comment]
> @kd-11 Awesome work! But is there a still a performance issue for NVIDIA or did you manage to find a workaround/fix? If not, I guess someone should open an issue to track that?

There is a workaround in place but performance will be worse on NV no matter what we do. There are over 6000 instances in some scenes and NV drivers don't like the inline updates very much.
We first need to get some real numbers on proper gaming GPUs and check if performance is unreasonable. If it's really bad then we can track it.

**Comment by RibShark** [comment]
This significantly harms performance for me on Ratchet and Clank: Tools of Destruction on my 3060 Ti, going from solid 60fps in Planet Sargasso to ~30 in a lot of areas. It would be great if there was a toggle for the old logic vs the new as I'd rather have a minor graphical issue with the foliage than play at low framerates.

**Comment by digant73** [comment]
> > This significantly harms performance for me on Ratchet and Clank: Tools of Destruction on my 3060 Ti, going from solid 60fps in Planet Sargasso to ~30 in a lot of areas. It would be great if there was a toggle for the old logic vs the new as I'd rather have a minor graphical issue with the foliage than play at low framerates.
> 
> The foliage did not render. It's not a graphical glitch, the emulator just skipped all the draw calls by accident due to a bug. It's really not acceptable to have such a thing remain in a mainline release.

Not understood if you refer to a regression caused by a PR or a limitation on NV drivers. In case of a regression on RPCS3, I suppose it should be fixable restoring/adapting the proper code after the broken PR is identified.

**Comment by kd-11** [comment]
> > > This significantly harms performance for me on Ratchet and Clank: Tools of Destruction on my 3060 Ti, going from solid 60fps in Planet Sargasso to ~30 in a lot of areas. It would be great if there was a toggle for the old logic vs the new as I'd rather have a minor graphical issue with the foliage than play at low framerates.
> > 
> > 
> > The foliage did not render. It's not a graphical glitch, the emulator just skipped all the draw calls by accident due to a bug. It's really not acceptable to have such a thing remain in a mainline release.
> 
> Not understood if you refer to a regression caused by a PR or a limitation on NV drivers. In case of a regression on RPCS3, I suppose it should be fixable restoring/adapting the proper code after the broken PR is identified.

It was broken since the emulator started, but the fix doesn't work well on nvidia due to driver limitations with vulkan 1.0. The performance drop is not a regression.

**Pull request merged**